### PR TITLE
fix: add time to the sdk.Context used in PrepareProposal

### PR DESCRIPTION
**Pull request by evan-forbes**
## Overview

I'm not sure if this fixes all of #2454, but it at least fixes one bug similar to it.

The fix in this PR allows for the antehandlers that are ran during prepare proposal to have access to a time close to the actual block time. It does this by simply adding `time.Now` to the header used for antehandler state access.

## Checklist

- [x] New and updated code has appropriate documentation
- [x] New and updated code has new and/or updated testing
- [x] Required CI checks are passing
- [x] Visual proof for any user facing features like CLI or documentation updates
- [x] Linked issues closed with keywords

### CODE REVIEW

File: app/test/prepare_proposal_context_test.go [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+	tmrand "github.com/tendermint/tendermint/libs/rand"
+)
+
+// TestTimeInPrepareProposalContext checks for an edge case where the block time
+// needs to be included in the sdk.Context that is being used in the
+// antehandlers. If a time is not included in the context, then the second
+// transaction in this test will always be filtered out, result in vesting
+// accounts never being able to spend funds.
+func TestTimeInPrepareProposalContext(t *testing.T) {
+	if testing.Short() {
```
Comment (evan-forbes) [inline app/test/prepare_proposal_context_test.go:29]: we could have included this in the standard cosmos-sdk test that we're already running, however it would be dependent upon the second testcase always being second

File: app/test/prepare_proposal_context_test.go [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+		t.Skip("skipping DebugVestingAccountsTime test in short mode.")
```
Comment (rootulp) [inline app/test/prepare_proposal_context_test.go:30]: [question] is this name supposed to match the test name?

```suggestion
		t.Skip("skipping TestTimeInPrepareProposalContext test in short mode.")
```

File: app/test/prepare_proposal_context_test.go [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+	type test struct {
+		name         string
+		msgFunc      func() (msgs []sdk.Msg, signer string)
+		expectedCode uint32
+	}
+	tests := []test{
+		{
+			name: "create continuous vesting account with a start time in the future",
+			msgFunc: func() (msgs []sdk.Msg, signer string) {
+				vestAccName := "vesting"
```
Comment (rootulp) [inline app/test/prepare_proposal_context_test.go:49]: [optional] I think this could be hoisted up to line ~38 and removed from this line and the second test case b/c IIUC the `vestAccName` should be identical for both test cases.

**Review by rootulp — DISMISSED** [review]

**Review by rootulp — DISMISSED** [review]

File: app/test/prepare_proposal_context_test.go (hunk 3 above)
Comment (evan-forbes) [reply app/test/prepare_proposal_context_test.go]: https://github.com/celestiaorg/celestia-app/pull/2515/commits/5ad837147e243a008735eb1faaa5679240a7b2c4

File: app/prepare_proposal.go [hunk 4]
```diff
@@ -18,7 +20,11 @@ import (
 func (app *App) PrepareProposal(req abci.RequestPrepareProposal) abci.ResponsePrepareProposal {
 	// create a context using a branch of the state and loaded using the
 	// proposal height and chain-id
-	sdkCtx := app.NewProposalContext(core.Header{ChainID: req.ChainId, Height: app.LastBlockHeight() + 1})
+	sdkCtx := app.NewProposalContext(core.Header{
+		ChainID: req.ChainId,
+		Height:  app.LastBlockHeight() + 1,
+		Time:    time.Now(),
```
Comment (cmwaters) [inline app/prepare_proposal.go:26]: There is a high possibility that time.Now() is greater than the block time proposed in the header (as it stems from the commit of the previous block). This could compromise the consistency property whereby a transaction seems valid in PrepareProposal but invalid in ProcessProposal. If the transaction isn't kicked from the mempool it could result in the chain halting (as proposers continue to propose a vesting transaction that they think is valid (it has a later time) but continues to get rejected (as validators are basing it off the last block time). Ideally we find a way to add the last block time to the context. If it's too difficult I would recommend setting Time to `time.Now().Sub(TargetBlockTime)` as an extra precaution but if there are multi rounds the same consistency problem could occur

**Review by cmwaters — CHANGES_REQUESTED** [review]

File: app/prepare_proposal.go (hunk 4 above)
Comment (evan-forbes) [reply app/prepare_proposal.go:26]: good point! I totally blanked this morning on us running the entire antehandler on all txs in process proposal

we definitely shouldn't do this

Comment (evan-forbes) [reply app/prepare_proposal.go:26]: I think the only solution here is to get the time in the request from comet similar to how its done on celestia-core/main or comet

Comment (liamsi) [inline app/prepare_proposal.go:26]: Yes, this must be the last block's time! Otherwise this is wrong. For example, when calculating the spendable amount, the only real reference the chain has is the last block, as the current block has not been finalized.

This looks about right:
https://github.com/cosmos/cosmos-sdk/blob/c7e0bd7b54d037a36ef7d31cc89ecabdf2124913/baseapp/abci.go#L409-L416

But please double-check that the `req` actually contains the correct time.

**Review by liamsi — CHANGES_REQUESTED** [review]
See either: https://github.com/celestiaorg/celestia-app/pull/2515/files#r1327951681
or: https://github.com/celestiaorg/celestia-app/pull/2515/files#r1327582804

File: app/prepare_proposal.go (hunk 4 above)
Comment (liamsi) [reply app/prepare_proposal.go:26]: Note: it does not seem to be the case that the correct time is set here:
https://github.com/celestiaorg/celestia-core/blob/d24c81afe0ecc4ccfbb7eb543a45a11794f04db4/state/execution.go#L118-L124

Only later in MakeBlock the time is set (correctly): 
- https://github.com/celestiaorg/celestia-core/blob/d24c81afe0ecc4ccfbb7eb543a45a11794f04db4/state/execution.go#L152
- https://github.com/celestiaorg/celestia-core/blob/3f3b7cc57f5cfc5e846ce781a9a407920e54fb72/state/state.go#L249

Comment (liamsi) [reply app/prepare_proposal.go:26]: Compare `main` of vanilla cometBFT: https://github.com/cometbft/cometbft/blob/1c6f86c840c64f83f4409ccca6b664234866cc38/state/execution.go#L128C9-L139

**Comment by evan-forbes** [comment]
Thanks again @cmwaters and @liamsi for catching the time.Now bug, I don't know what I was thinking. Too many late nights deugging testground last week! :sweat_smile: . Per the first and second thread, the only solution to this is to use the same time that will be used in process proposal (the time in the header from the last commit). That's what's done in https://github.com/celestiaorg/celestia-core/pull/1081, which should be merged before this PR.

File: app/test/prepare_proposal_context_test.go [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+				msg := banktypes.NewMsgSend(
+					vestAccAddr,
+					sendingAccAddr,
+					sdk.NewCoins(sdk.NewCoin(app.BondDenom, sdk.NewInt(1))),
+				)
+				return []sdk.Msg{msg}, vestAccName
+			},
+			expectedCode: abci.CodeTypeOK,
+		},
+	}
```
Comment (evan-forbes) [inline app/test/prepare_proposal_context_test.go:81]: one additional test we can do is to have a third test case where we attempt to send too much and expect it to fail. I have a commit with this ready, however I'm worried that it will cause flakiness do to the process requiring a specific time delay window.

**Review by rootulp — APPROVED** [review]

**Review by liamsi — APPROVED** [review]
👏🏼

File: app/test/prepare_proposal_context_test.go [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+	abci "github.com/tendermint/tendermint/abci/types"
+	tmrand "github.com/tendermint/tendermint/libs/rand"
+)
+
+// TestTimeInPrepareProposalContext checks for an edge case where the block time
+// needs to be included in the sdk.Context that is being used in the
+// antehandlers. If a time is not included in the context, then the second
+// transaction in this test will always be filtered out, result in vesting
+// accounts never being able to spend funds.
+func TestTimeInPrepareProposalContext(t *testing.T) {
```
Comment (cmwaters) [inline app/test/prepare_proposal_context_test.go:28]: I feel like we defer to using integration tests too often when unit tests would suffice

File: app/test/prepare_proposal_context_test.go [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+				_, _, err := cctx.Keyring.NewMnemonic(vestAccName, keyring.English, "", "", hd.Secp256k1)
+				require.NoError(t, err)
+				sendAcc := accounts[0]
+				sendingAccAddr := testfactory.GetAddress(cctx.Keyring, sendAcc)
+				vestAccAddr := testfactory.GetAddress(cctx.Keyring, vestAccName)
+				msg := vestingtypes.NewMsgCreateVestingAccount(
+					sendingAccAddr,
+					vestAccAddr,
+					sdk.NewCoins(sdk.NewCoin(app.BondDenom, sdk.NewInt(1000000))),
+					time.Now().Unix(),
```
Comment (cmwaters) [inline app/test/prepare_proposal_context_test.go:58]: This is not a start time in the future as is stated in the test name

File: app/test/prepare_proposal_context_test.go (hunk 5 above)
Comment (cmwaters) [reply app/test/prepare_proposal_context_test.go:81]: I don't think it will be flaky if you set the end time well in the future and try send all the funds

**Review by cmwaters — APPROVED** [review]
Few minor comments but mostly looks good

File: app/test/prepare_proposal_context_test.go (hunk 5 above)
Comment (evan-forbes) [reply app/test/prepare_proposal_context_test.go:81]: > well in the future and try send all the funds

yeah good point, sorry I was not clear enough. the thing that I was trying to test with the additional case here was not just that a tx would fail if there was not enough of a spendable balance, I think we have that somehwere already (#2004 iirc). That wouldn't be flakey. 

I was trying to create a test that checks that puts us in the exact scenario we would get in if we used `time.Now()`. Where a tx doesn't get filtered but is invalid.

File: app/test/prepare_proposal_context_test.go (hunk 6 above)
Comment (evan-forbes) [reply app/test/prepare_proposal_context_test.go:28]: https://github.com/celestiaorg/celestia-app/issues/2533

File: app/test/prepare_proposal_context_test.go (hunk 7 above)
Comment (evan-forbes) [reply app/test/prepare_proposal_context_test.go:58]: will grab this in a followup
